### PR TITLE
Damage Components Attribute

### LINKED ISSUE #145: Option to damage specific components of vehicles
**Is your feature request related to a problem?**
No actual problem, just a nice feature I've gotten used to. Our unit uses it to test pilots for helicopter failures, or to pop tires to start a convoy ambush.

**Solution you'd like:**
UI menu similar to Achilles's layout. Advanced damage tab that brings a list of all the components of the selected vehicle, with the ability to set the damage value of each one.

**Alternatives you've considered:**
Only other solution would be to use scripts and triggers in Eden

**Additional context:**
Just want to say you guys are doing a great job.

### PR DESCRIPTION
**Pull request by neilzar**
**When merged this pull request will:**
- Close #145

### CODE REVIEW

**Comment by neilzar** [comment]
Only applies to a single vehicle rn, trying to figure out the best way to filter for the same vehicle class in selection.

File: addons/attributes/XEH_postInit.sqf [hunk 1]
```diff
@@ -26,3 +26,11 @@
         "reloadSpeed"
     ];
 }] call CBA_fnc_addEventHandler;
+
+[QGVAR(setAllHitPointsDamage), {
```
Comment (CreepPork) [inline addons/attributes/XEH_postInit.sqf]: Perhaps renamed to just `setHitPointsDamage`?

**Comment by CreepPork** [comment]
Also might have to postpone this after #140.

File: addons/attributes/XEH_postInit.sqf (hunk 1 above)
Comment (neilzar) [reply addons/attributes/XEH_postInit.sqf]: Avoided that because of https://community.bistudio.com/wiki/setHitPointDamage

**Comment by CreepPork** [comment]
Also, could you attach a screenshot, please?

**Comment by neilzar** [comment]
image
You could see it when you test it in-game.

File: addons/common/functions/fnc_getHitPointString.sqf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        _text = localize _combinedString;
+    };
+
+    // Cut off one character
+    _toFind = [_toFind, 0, count _toFind - 1] call CBA_fnc_substr;
+};
+
+// Don't display part name if no string is found in stringtable
+if (_text == LSTRING(Hit)) then {
+    if (_hitPoint != "") then { LOG_1("Hitpoint [%1] - could not be localized", _hitPoint); };
```
Comment (CreepPork) [inline addons/common/functions/fnc_getHitPointString.sqf]: Is the log necessary and this is quite a long line could push down one line

File: addons/common/functions/fnc_getHitPointString.sqf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Finds the localized string of the given hitpoint name or uses the hitpoint name if none found.
+ *
+ * Arguments:
+ * Hitpoint <STRING>
+ *
+ * Return Value:
+ * Localized Hit Point Name <STRING>
+ *
+ * Example:
+ * "HitFuel" call zen_common_fnc_getHitPointString
```
Comment (CreepPork) [inline addons/common/functions/fnc_getHitPointString.sqf]:
```suggestion
 * ["HitFuel"] call zen_common_fnc_getHitPointString
```

File: addons/common/functions/fnc_getHitPointString.sqf (hunk 2 above)
Comment (neilzar) [reply addons/common/functions/fnc_getHitPointString.sqf]: I removed it as it was leftover from testing.

File: addons/attributes/functions/fnc_buttonComponents.sqf [hunk 4]
```diff
@@ -0,0 +1,39 @@
+#include "script_component.hpp"
+/*
+ * Author: mharis001
```
Comment (mharis001) [inline addons/attributes/functions/fnc_buttonComponents.sqf]: I did not author this.

File: addons/attributes/functions/fnc_buttonComponents.sqf [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * Example:
+ * [] call zen_attributes_fnc_buttonComponents
+ *
+ * Public: No
+ */
+
+private _vehicle = GETMVAR(BIS_fnc_initCuratorAttributes_target,objNull);
+private _components = [];
+(getAllHitPointsDamage _vehicle) params ["_allHitPointNames", "_allHitSelectionNames", "_allHitPointsDamage"];
```
Comment (mharis001) [inline addons/attributes/functions/fnc_buttonComponents.sqf]: Don't need the parentheses, and for variable names - I think: `["_hitPointNames", "_selectionNames", "_damageValues"]` might be better.

File: addons/attributes/functions/fnc_buttonComponents.sqf [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    private _component = _x;
+    if (_component isEqualTo "") then { _component = _allHitSelectionNames select _forEachIndex };
+    private _componentName = _component call EFUNC(common,getHitPointString);
+
+    _components pushBack ["SLIDER:PERCENT", [_componentName, _component], [0, 1, _allHitPointsDamage select _forEachIndex], true];
+} forEach _allHitPointNames;
+
+[LSTRING(ComponentsDamage), _components, {
+    params ["_dialogValues", "_vehicle"];
+    private _vehicleType = typeOf _vehicle;
```
Comment (mharis001) [inline addons/attributes/functions/fnc_buttonComponents.sqf]: Blank line after `params`.

File: addons/attributes/functions/fnc_buttonComponents.sqf [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ * Public: No
+ */
+
+private _vehicle = GETMVAR(BIS_fnc_initCuratorAttributes_target,objNull);
+private _components = [];
+(getAllHitPointsDamage _vehicle) params ["_allHitPointNames", "_allHitSelectionNames", "_allHitPointsDamage"];
+
+{
+    private _component = _x;
+    if (_component isEqualTo "") then { _component = _allHitSelectionNames select _forEachIndex };
```
Comment (mharis001) [inline addons/attributes/functions/fnc_buttonComponents.sqf]: Multi-line and some spacing might look better.

File: addons/attributes/functions/fnc_buttonComponents.sqf (hunk 5 above)
Comment (CreepPork) [reply addons/attributes/functions/fnc_buttonComponents.sqf]: But it's clearer with the parentheses. I'd agree with the variables though.

Comment (neilzar) [reply addons/attributes/functions/fnc_buttonComponents.sqf]: I agree with Creep that the parentheses improve readability, so I have kept those.

**Comment by mharis001** [comment]
Will need some adjustments after attributes rewrite goes through, should merge after that.

**Pull request merged**